File: Assets/ProjectGame/Lib/data/enum/Sound_UI_Title/Sound_UI_TitleID.py
from enum import IntEnum
from typing import Callable, List, Any


class Sound_UI_TitleID(IntEnum):
    NONE = 0  # デフォルト値（C#のNoneに相当）
    SelectMove = 1  # 移動音
    PushEnter = 2  # 決定音
    MAX = 3

def to_int(id: "Sound_UI_TitleID") -> int:
    return int(id)


def to_Sound_UI_TitleID(id: int) -> "Sound_UI_TitleID":
    return Sound_UI_TitleID(id)


def to_index(id: "Sound_UI_TitleID") -> int:
    return int(id) - 1


def for_id(action: Callable[["Sound_UI_TitleID"], None]):
    if action is None:
        raise ValueError("action cannot be None")
    start = Sound_UI_TitleID.SelectMove.value
    for i in range(start, Sound_UI_TitleID.MAX.value):
        try:
            value = Sound_UI_TitleID(i)
            action(value)
        except ValueError:
            continue  # 未定義の値はスキップ


def find_all(predicate: Callable[["Sound_UI_TitleID"], bool]) -> List["Sound_UI_TitleID"]:
    if predicate is None:
        raise ValueError("predicate cannot be None")
    results: List["Sound_UI_TitleID"] = []
    start = Sound_UI_TitleID.SelectMove.value
    for i in range(start, Sound_UI_TitleID.MAX.value):
        try:
            value = Sound_UI_TitleID(i)
            if predicate(value):
                results.append(value)
        except ValueError:
            continue
    return results


def find(predicate: Callable[["Sound_UI_TitleID"], bool]) -> "Sound_UI_TitleID":
    if predicate is None:
        raise ValueError("predicate cannot be None")
    start = Sound_UI_TitleID.SelectMove.value
    for i in range(start, Sound_UI_TitleID.MAX.value):
        try:
            value = Sound_UI_TitleID(i)
            if predicate(value):
                return value
        except ValueError:
            continue
    return Sound_UI_TitleID.NONE

File: Assets/ProjectGame/Lib/data/enum/Sound_UI_Title/Sound_UI_TitleID.js
// Sound_UI_TitleID - Pure JavaScript Version
export const Sound_UI_TitleID = {
    None: 0,  // デフォルト値（C#互換）
    SelectMove: 1,  // 移動音
    PushEnter: 2,  // 決定音
    Max: 3
};

export const Sound_UI_TitleIDExtensions = {
    /**
     * Enumを数値に変換
     */
    toInt(id) {
        return Number(id);
    },

    /**
     * 数値をEnumに変換
     */
    toSound_UI_TitleID(id) {
        return id;
    },

    /**
     * 0-based indexに変換
     */
    toIndex(id) {
        return Number(id) - 1;
    },

    /**
     * すべてのIDに対して処理を実行
     */
    forID(action) {
        if (typeof action !== 'function') {
            throw new Error('action must be a function');
        }
        const start = Sound_UI_TitleID.SelectMove;
        const max = Sound_UI_TitleID.Max;
        for (let id = start; id < max; id++) {
            if (Object.values(Sound_UI_TitleID).includes(id)) {
                action(id);
            }
        }
    },

    /**
     * 条件に合うすべてのIDを返す
     */
    findAll(predicate) {
        if (typeof predicate !== 'function') {
            throw new Error('predicate must be a function');
        }
        const results = [];
        const start = Sound_UI_TitleID.SelectMove;
        const max = Sound_UI_TitleID.Max;
        for (let id = start; id < max; id++) {
            if (Object.values(Sound_UI_TitleID).includes(id)) {
                if (predicate(id)) {
                    results.push(id);
                }
            }
        }
        return results;
    },

    /**
     * 条件に合う最初のIDを返す（見つからなければ None）
     */
    find(predicate) {
        if (typeof predicate !== 'function') {
            throw new Error('predicate must be a function');
        }
        const start = Sound_UI_TitleID.SelectMove;
        const max = Sound_UI_TitleID.Max;
        for (let id = start; id < max; id++) {
            if (Object.values(Sound_UI_TitleID).includes(id)) {
                if (predicate(id)) {
                    return id;
                }
            }
        }
        return Sound_UI_TitleID.None;
    }
};
